vllm/v1/core/specialized_manager.py

```diff
@@ -0,0 +1,175 @@
+# SPDX-License-Identifier: Apache-2.0
+from abc import ABC, abstractmethod
+from itertools import chain
+
+from vllm.v1.core.block_pool import BlockPool
+from vllm.v1.core.kv_cache_utils import (BlockHashType, KVCacheBlock,
+                                         PrefixLengthRange)
+from vllm.v1.kv_cache_interface import (FullAttentionSpec, KVCacheSpec,
+                                        SlidingWindowSpec)
+
+
+class SpecializedManager(ABC):
+    """
+    An abstract base class for specialized managers that handle the kv
+    cache management logic of different attention layers.
+    """
+
+    def __init__(
+        self,
+        kv_cache_spec: KVCacheSpec,
+        block_pool: BlockPool,
+    ) -> None:
+        """
+        Initializes the SpecializedManager.
+        Args:
+            kv_cache_spec: The kv_cache_spec for this manager.
+            block_pool: The block pool.
+        """
+
+        self.block_size = kv_cache_spec.block_size
+        self.kv_cache_spec = kv_cache_spec
+        self.block_pool = block_pool
+
+    @abstractmethod
+    def get_possible_cached_prefix(
+        self, block_hashes: list[BlockHashType]
+    ) -> tuple[list[PrefixLengthRange], list[KVCacheBlock]]:
+        """
+        Get the possible cached prefixes of a request based on its block hashes.
+        If no cached prefixes are found, returns a tuple with a prefix length 
+        range of [0, 0] and an empty list of blocks.
+
+        Args:
+            block_hashes: The block hashes of the request.
+        Returns:
+            A tuple containing:
+                - A list of all possible cached prefix lengths.
+                - A list of cached blocks for each block hash. Use the null 
+                block for blocks that are not cached. Can skip the last blocks
+                that are not cached.
+        """
+
+        raise NotImplementedError
+
+    @abstractmethod
+    def remove_useless_blocks(self, block_table: list[KVCacheBlock],
```

Comment:
Shouldn't we call it `blocks` of `kv_blocks` rather than block table?
```suggestion
    def remove_useless_blocks(self, blocks: list[KVCacheBlock],
```